### PR TITLE
[SPARK-47416][SQL] Add new functions to CollationBenchmark

### PR DESCRIPTION
**Pull request by vladimirg-db**
### What changes were proposed in this pull request?
Added new benchmarks for contains, startsWith, endsWith prior to improving the implementation for the UTF8_BINARY_LCASE collation.

### Why are the changes needed?
To see exact improvements after the implementation of https://issues.apache.org/jira/browse/SPARK-47418

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
GHA 'Run Benchmarks' ran on this, for both JDK 17 and JDK 21

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala [hunk 1]
```diff
@@ -100,6 +100,90 @@ abstract class CollationBenchmarkBase extends BenchmarkBase {
...
+      collationTypes: Seq[String],
+      utf8Strings: Seq[UTF8String]): Unit = {
+    val sublistStrings = utf8Strings
+
+    val benchmark = new Benchmark(
+      "collation unit benchmarks - contains",
+      utf8Strings.size * 10,
+      warmupTime = 10.seconds,
+      output = output)
+    collationTypes.foreach(collationType => {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:114]: can you use `foreach { collationType =>` instead

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala [hunk 2]
```diff
@@ -85,19 +86,103 @@ abstract class CollationBenchmarkBase extends BenchmarkBase {
...
-          )
-        )
+            }
+          }
+        }
+      }
+    }
+    }
+    benchmark.run()
+  }
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:103]: we shouldn't make unnecessary changes

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala [hunk 3]
```diff
@@ -36,18 +36,19 @@ abstract class CollationBenchmarkBase extends BenchmarkBase {
...
+          utf8Strings.foreach { s =>
+            (0 to 10).foreach { _ =>
+              collation.equalsFunction(s, s1).booleanValue()
+            }
+          }
+        }
       }
     }
-    )
+    }
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:51]: we shouldn't make unnecessary changes

File: sql/core/benchmarks/CollationBenchmark-jdk21-results.txt [hunk 4]
```diff
@@ -2,26 +2,53 @@ OpenJDK 64-Bit Server VM 21.0.2+13-LTS on Linux 6.5.0-1017-azure
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 11716          11716           1          0.0      117157.9       1.0X
-UNICODE                                          180133         180137           5          0.0     1801332.1       0.1X
-UTF8_BINARY                                       10476          10477           1          0.0      104757.4       1.1X
-UNICODE_CI                                       148171         148190          28          0.0     1481705.6       0.1X
+UTF8_BINARY_LCASE                                 11380          11390          15          0.0      113795.1       1.0X
+UNICODE                                          176471         176477           9          0.0     1764709.9       0.1X
+UTF8_BINARY                                       10858          10859           1          0.0      108580.4       1.0X
+UNICODE_CI                                       153060         153161         143          0.0     1530598.1       0.1X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationBenchmark-jdk21-results.txt:26]: I see that the benchmark runs on previous functions too, but we shouldn't make any unnecessary modifications in this PR (that are not related to the changes brought by this PR)

so I would advise keep the previous results as they were, and only append yours below

File: sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt [hunk 5]
```diff
@@ -2,26 +2,53 @@ OpenJDK 64-Bit Server VM 21.0.2+13-LTS on Linux 6.5.0-1017-azure
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                  9183           9230          67          0.0      229564.0       1.0X
-UNICODE                                           38937          38952          22          0.0      973421.3       0.2X
-UTF8_BINARY                                        1376           1376           0          0.0       34397.5       6.7X
-UNICODE_CI                                        32881          32882           1          0.0      822027.4       0.3X
+UTF8_BINARY_LCASE                                 10768          10780          17          0.0      269199.8       1.0X
+UNICODE                                           42046          42062          23          0.0     1051138.9       0.3X
+UTF8_BINARY                                        1980           1984           5          0.0       49506.4       5.4X
+UNICODE_CI                                        33206          33212           7          0.0      830160.4       0.3X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationNonASCIIBenchmark-jdk21-results.txt:26]: I see that the benchmark runs on previous functions too, but we shouldn't make any unnecessary modifications in this PR (that are not related to the changes brought by this PR)

so I would advise keep the previous results as they were, and only append yours below

File: sql/core/benchmarks/CollationNonASCIIBenchmark-results.txt [hunk 6]
```diff
@@ -2,26 +2,53 @@ OpenJDK 64-Bit Server VM 17.0.10+7-LTS on Linux 6.5.0-1017-azure
...
 ------------------------------------------------------------------------------------------------------------------------
-UTF8_BINARY_LCASE                                 10494          10499           6          0.0      262360.0       1.0X
-UNICODE                                           40410          40422          17          0.0     1010261.8       0.3X
-UTF8_BINARY                                        2035           2035           1          0.0       50877.8       5.2X
-UNICODE_CI                                        31470          31493          32          0.0      786752.4       0.3X
+UTF8_BINARY_LCASE                                 10002          10021          27          0.0      250062.0       1.0X
+UNICODE                                           39592          39623          45          0.0      989789.3       0.3X
+UTF8_BINARY                                        2029           2032           4          0.0       50718.3       4.9X
+UNICODE_CI                                        33172          33174           2          0.0      829309.4       0.3X
```
Comment (uros-db) [inline sql/core/benchmarks/CollationNonASCIIBenchmark-results.txt:26]: I see that the benchmark runs on previous functions too, but we shouldn't make any unnecessary modifications in this PR (that are not related to the changes brought by this PR)

so I would advise keep the previous results as they were, and only append yours below

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala [hunk 7]
```diff
@@ -85,19 +86,103 @@ abstract class CollationBenchmarkBase extends BenchmarkBase {
...
+              )
+            }
+          }
+        }
       }
     }
-    )
+    }
+    benchmark.run()
+  }
```
Comment (uros-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:131]: is there a need to switch to braces?

if not, I would say we should try to stay consistent with the formatting - if the previous version used parentheses, then we should keep everything as it is

(ditto below)

File: sql/core/benchmarks/CollationNonASCIIBenchmark-results.txt (hunk 6 above)
Comment (vladimirg-db) [reply sql/core/benchmarks/CollationNonASCIIBenchmark-results.txt:26]: I agree with you on that. But since those files are auto-generated it makes sense to store them like they were produced. However, this is a trivial change so I will revert those to previous values.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala (hunk 7 above)
Comment (vladimirg-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:131]: This was a request of [HyukjinKwon](https://github.com/HyukjinKwon) (see his comment above)

**Review by uros-db — CHANGES_REQUESTED** [review]
some small changes recommended, otherwise looks great!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala (hunk 7 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:131]: sorry, didn't catch that - my bad!

go ahead and keep these { } changes then

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala [hunk 8]
```diff
@@ -130,6 +215,9 @@ object CollationBenchmark extends CollationBenchmarkBase {
     benchmarkUTFStringEquals(collationTypes, generateSeqInput(10000L))
     benchmarkUTFStringCompare(collationTypes, generateSeqInput(10000L))
     benchmarkUTFStringHashFunction(collationTypes, generateSeqInput(10000L))
+    benchmarkUTFStringContains(collationTypes, generateSeqInput(10000L))
+    benchmarkUTFStringStartsWith(collationTypes, generateSeqInput(10000L))
+    benchmarkUTFStringEndsWith(collationTypes, generateSeqInput(10000L))
```
[the remark this answers is not in the document]
Comment (vladimirg-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala:220]: I would just name those `benchmarkContains `, `benchmarkStartsWith`, `benchmarkEndsWith`, since here we are in the context of CollationBenchmark anyways.

**Review by uros-db — APPROVED** [review]
LGTM @cloud-fan @HyukjinKwon please review

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!
